### PR TITLE
SAMSON-193 Fix child processes having access to all of samsons env, including se…

### PR DESCRIPTION
**Pull request by grosser**
https://zendesk.atlassian.net/browse/SAMSON-193

Also fixes race condition where with_clean_env might have just ended in another thread and even .env values get exposed

@jonmoter @steved @dragonfax 

/cc @zendesk/samson 

### Risks
 - High: all kinds of deploys failing

### CODE REVIEW

File: app/models/terminal_executor.rb [hunk 1]
```diff
@@ -34,26 +33,26 @@ def stop!(signal)
...
-    _, status = Process.wait2(pid)
-
-    input.close
+      wait_thr.value
+    end.success?
+  end
 
-    return status.success?
+  def whitelisted_env
+    whitelist = ['TERM', 'SHELL'] + ENV.keys.select { |k| k.end_with?('PATH', 'DIR', '_HOME') }
```
Comment (jonmoter) [inline app/models/terminal_executor.rb]: Seems kind of arbitrary that we're keeping env vars that end with `PATH`, `DIR`, or `_HOME`.

I would rather have this whitelist be minimal and documented, with only the ENV vars that would be necessary for the commands to function.

**Comment by jonmoter** [comment]
Thanks for finding a fix for this! It's been bugging me for a while now.

File: app/models/terminal_executor.rb (hunk 1 above)
Comment (grosser) [inline app/models/terminal_executor.rb]: we need path / java_home / tempdir / gem_home / gem_path ... possibly gopath / nodpath etc ... so I was thinking the pattern here is all paths are fine ... either way is fine I guess, just means a few fix-up PRs will be coming soonish :)
so which one do you like more, all paths / exact whitelist ?

**Comment by grosser** [comment]
updated to reduced whitelist + ENV_WHITELIST support

File: test/models/terminal_executor_test.rb [hunk 2]
```diff
@@ -1,6 +1,13 @@
 require_relative '../test_helper'
 
 describe TerminalExecutor do
+  def with_env(k, v)
```
Comment (jonmoter) [inline test/models/terminal_executor_test.rb]: How about having this take a hash of new values, rather than a single k/v pair?

Comment (grosser) [inline test/models/terminal_executor_test.rb]: updates

**Comment by jonmoter** [comment]
:+1:

**Pull request merged**